clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+    private SaslClient saslClient;
+    private Subject subject;
+    private String servicePrincipal;
+    private PrincipalBuilder principalBuilder;
+    private String host;
+    private String node;
+    private String serverRealm;
+    private Object[] principals;
+    private Principal clientPrincipal;
+    private KerberosName clientKerberosName;
+    private KerberosName serviceKerberosName;
+    private String clientPrincipalName;
+    private TransportLayer transportLayer;
+    private NetworkReceive netInBuffer;
+    private NetworkSend netOutBuffer;
+    private byte[] saslToken = new byte[0];
+
+    public enum SaslState {
+        INITIAL, INTERMEDIATE, COMPLETE, FAILED
+    }
+
+    private SaslState saslState = SaslState.INITIAL;
+
+    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host) throws IOException {
+        this.node = node;
+        this.subject = subject;
+        this.host = host;
+        this.servicePrincipal = servicePrincipal;
+    }
+
+    public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) throws KafkaException {
+        try {
+            this.transportLayer = transportLayer;
+            this.principalBuilder = principalBuilder;
+
+            this.principals = subject.getPrincipals().toArray();
+            // determine client principal from subject.
+            this.clientPrincipal = (Principal) principals[0];
+            this.clientKerberosName = new KerberosName(clientPrincipal.getName());
+            // assume that server and client are in the same realm (by default; unless the config
+            // "kafka.server.realm" is set).
+            if (configs.containsKey(SaslConfigs.SASL_KAFKA_SERVER_REALM))
+                this.serverRealm = (String) configs.get(SaslConfigs.SASL_KAFKA_SERVER_REALM);
+            else
+                this.serverRealm = clientKerberosName.realm();
+            this.serviceKerberosName = new KerberosName(servicePrincipal + "@" + serverRealm);
+            this.clientPrincipalName = clientKerberosName.toString();
+            this.saslClient = createSaslClient();
+        } catch (Exception e) {
+            throw new KafkaException("Failed to configure SaslClientAuthenticator", e);
+        }
+    }
+
+    private SaslClient createSaslClient() {
+        try {
+            saslClient = Subject.doAs(subject, new PrivilegedExceptionAction<SaslClient>() {
+                    public SaslClient run() throws SaslException {
+                        LOG.debug("Client will use GSSAPI as SASL mechanism.");
+                        String[] mechs = {"GSSAPI"};
+                        LOG.debug("creating sasl client: client=" + clientPrincipalName + ";service=" + servicePrincipal + ";serviceHostname=" + host);
+                        SaslClient saslClient = Sasl.createSaslClient(mechs, clientPrincipalName, servicePrincipal, host, null, new ClientCallbackHandler(null));
```

Comment:
didn't understand , are you saying we should pass jaas config file as part of client config properties?